File: bindings/android/OpenEyes.kt
package com.cambriantech.openeyes

import android.media.Image
import java.nio.ByteBuffer

/**
 * OpenEyes — Kotlin wrapper for the Rust open-eyes engine.
 *
 * Thin binding. All compute is in Rust. This file handles:
 * 1. Camera frame delivery (CameraX, Camera2, Meta Quest Passthrough)
 * 2. AR pose delivery (ARCore, OVR SDK)
 * 3. Event polling back to the caller
 *
 * Works on: Android phones/tablets, Meta Quest (2/3/Pro), Android Auto cameras
 * The engine doesn't care about the platform — it receives frames + poses.
 */
class OpenEyesEngine : AutoCloseable {

    // Native engine pointer (opaque, managed by Rust)
    private var enginePtr: Long = 0L

    /** Motion magnitude (0.0 = static, higher = more motion) */
    var motionMagnitude: Float = 0f
        private set

    /** Total frames processed */
    val frameCount: Long get() = if (enginePtr != 0L) nativeGetFrameCount(enginePtr) else 0L

    init {
        System.loadLibrary("openeyes")
        enginePtr = nativeCreate()
    }

    override fun close() {
        if (enginePtr != 0L) {
            nativeDestroy(enginePtr)
            enginePtr = 0L
        }
    }

    // MARK: - Frame input

    /**
     * Push a raw camera frame from CameraX ImageAnalysis or Camera2 ImageReader.
     *
     * HOT PATH. The YUV planes go straight to Rust via JNI. No Java-side pixel processing.
     *
     * On Meta Quest: same path. Passthrough API delivers frames as Image objects
     * with the same YUV_420_888 format.
     */
    fun pushFrame(cameraId: Int, image: Image, rotation: Int = 0) {
        if (enginePtr == 0L) return

        val width = image.width
        val height = image.height

        when (image.format) {
            android.graphics.ImageFormat.YUV_420_888 -> {
                // YUV420 — the standard Android camera output.
                // Pack Y + UV planes into a single buffer for the FFI.
                val yPlane = image.planes[0].buffer
                val uPlane = image.planes[1].buffer
                val vPlane = image.planes[2].buffer

                val ySize = yPlane.remaining()
                val uSize = uPlane.remaining()
                val vSize = vPlane.remaining()

                // Allocate once, reuse via ThreadLocal in production
                val packed = ByteBuffer.allocateDirect(ySize + uSize + vSize)
                packed.put(yPlane)
                packed.put(uPlane)
                packed.put(vPlane)
                packed.rewind()

                nativePushFrame(
                    enginePtr, cameraId,
                    packed, packed.remaining(),
                    width, height,
                    PIXEL_FORMAT_YUV420, rotation
                )
            }
            android.graphics.ImageFormat.FLEX_RGBA_8888 -> {
                val buffer = image.planes[0].buffer
                nativePushFrame(
                    enginePtr, cameraId,
                    buffer, buffer.remaining(),
                    width, height,
                    PIXEL_FORMAT_RGB8, rotation
                )
            }
        }

        motionMagnitude = nativeGetMotion(enginePtr)
    }

    /**
     * Push a raw byte buffer directly. For non-Image sources (screenshots, RTSP decoded frames).
     */
    fun pushFrameBuffer(
        cameraId: Int,
        buffer: ByteBuffer,
        width: Int, height: Int,
        format: Int,
        rotation: Int = 0
    ) {
        if (enginePtr == 0L) return
        nativePushFrame(enginePtr, cameraId, buffer, buffer.remaining(), width, height, format, rotation)
        motionMagnitude = nativeGetMotion(enginePtr)
    }

    // MARK: - AR Pose input

    /**
     * Set camera pose from ARCore (Pose / Camera.getDisplayOrientedPose()).
     * Column-major 4x4 float matrix — 16 floats.
     *
     * On Meta Quest: OVR SDK provides headset pose as a 4x4 matrix.
     * Same call, same format.
     */
    fun setCameraPose(cameraId: Int, transform: FloatArray) {
        if (enginePtr == 0L || transform.size != 16) return
        nativeSetCameraPose(enginePtr, cameraId, transform)
    }

    // MARK: - Camera intrinsics

    /**
     * Set camera intrinsics. Call once per camera.
     * On ARCore: extract from CameraIntrinsics.getFocalLength() / getPrincipalPoint()
     */
    fun setIntrinsics(
        cameraId: Int,
        fx: Double, fy: Double,
        cx: Double, cy: Double,
        width: Int, height: Int,
        k1: Double = 0.0, k2: Double = 0.0, k3: Double = 0.0
    ) {
        if (enginePtr == 0L) return
        nativeSetIntrinsics(enginePtr, cameraId, fx, fy, cx, cy, width, height, k1, k2, k3)
    }

    // MARK: - JNI native methods

    private external fun nativeCreate(): Long
    private external fun nativeDestroy(engine: Long)
    private external fun nativePushFrame(
        engine: Long, cameraId: Int,
        data: ByteBuffer, len: Int,
        width: Int, height: Int,
        format: Int, rotation: Int
    )
    private external fun nativeSetCameraPose(engine: Long, cameraId: Int, transform: FloatArray)
    private external fun nativeGetMotion(engine: Long): Float
    private external fun nativeGetFrameCount(engine: Long): Long
    private external fun nativeSetIntrinsics(
        engine: Long, cameraId: Int,
        fx: Double, fy: Double,
        cx: Double, cy: Double,
        width: Int, height: Int,
        k1: Double, k2: Double, k3: Double
    )

    companion object {
        const val PIXEL_FORMAT_BGRA8 = 0
        const val PIXEL_FORMAT_YUV420 = 1
        const val PIXEL_FORMAT_RGB8 = 2
        const val PIXEL_FORMAT_GRAY8 = 3
    }
}

/**
 * Abstract camera source — the engine consumes frames from any source.
 *
 * Known implementations:
 * - PhoneCameraSource: CameraX ImageAnalysis (phones/tablets)
 * - ARCameraSource: ARCore Session (AR-capable devices)
 * - QuestCameraSource: Passthrough API (Meta Quest headsets)
 * - RtspCameraSource: FFmpeg-decoded RTSP stream (IP cameras)
 */
interface OpenEyesCameraSource {
    val cameraId: Int
    fun start()
    fun stop()
}

// MARK: - CameraX camera source

// In a real implementation this would be:
//
// class PhoneCameraSource(
//     override val cameraId: Int,
//     private val engine: OpenEyesEngine,
//     private val lifecycleOwner: LifecycleOwner
// ) : OpenEyesCameraSource {
//
//     private val cameraProviderFuture = ProcessCameraProvider.getInstance(context)
//
//     override fun start() {
//         val imageAnalysis = ImageAnalysis.Builder()
//             .setTargetResolution(Size(1280, 720))
//             .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
//             .build()
//
//         imageAnalysis.setAnalyzer(executor) { imageProxy ->
//             // HOT PATH: image → Rust, no Java processing
//             imageProxy.image?.let { engine.pushFrame(cameraId, it) }
//             imageProxy.close()
//         }
//
//         cameraProvider.bindToLifecycle(lifecycleOwner, cameraSelector, imageAnalysis, preview)
//     }
// }
//
// QuestCameraSource follows the same pattern via OVR Passthrough API.
// The frame delivery contract is identical: Image → pushFrame → Rust.

File: bindings/ios/OpenEyes.swift
/// OpenEyes — Swift wrapper for the Rust open-eyes engine.
///
/// Thin binding. All compute is in Rust. This file handles:
/// 1. Camera frame delivery (AVFoundation, ARKit, visionOS passthrough)
/// 2. AR pose delivery (ARKit / ARSession)
/// 3. Event polling back to the caller
///
/// Works on: iOS, iPadOS, visionOS (Vision Pro), macOS
/// The engine doesn't care about the platform — it receives frames + poses.
/// Camera source abstraction handles the API differences.

import Foundation
import AVFoundation

// MARK: - C FFI imports (from openeyes.h)

// These are the raw C function pointers from libopeneyes.
// In a real build, this comes from the bridging header or module map.
// Declaring them here for documentation — the actual linkage is via
// the .xcframework's module.modulemap.

/*
 func oe_create() -> OpaquePointer
 func oe_destroy(_ engine: OpaquePointer)
 func oe_push_frame(_ engine: OpaquePointer, _ camera_id: UInt32,
                     _ data: UnsafePointer<UInt8>, _ len: Int,
                     _ width: UInt32, _ height: UInt32,
                     _ format: UInt32, _ rotation: Int32) -> Int32
 func oe_set_camera_pose(_ engine: OpaquePointer, _ camera_id: UInt32,
                          _ transform: UnsafePointer<Float>) -> Int32
 func oe_get_motion(_ engine: OpaquePointer) -> Float
 func oe_get_frame_count(_ engine: OpaquePointer) -> UInt64
 func oe_set_intrinsics(_ engine: OpaquePointer, _ camera_id: UInt32,
                         _ fx: Double, _ fy: Double,
                         _ cx: Double, _ cy: Double,
                         _ width: UInt32, _ height: UInt32,
                         _ k1: Double, _ k2: Double, _ k3: Double) -> Int32
 func oe_poll_events(_ engine: OpaquePointer,
                      _ callback: @convention(c) (UnsafePointer<OEEvent>?) -> Void) -> Int32
*/

// MARK: - Pixel format constants (must match OEPixelFormat in Rust)

public enum OEPixelFormat: UInt32 {
    case bgra8 = 0    // iOS AVCaptureSession default
    case yuv420 = 1   // Android Camera2 / some iOS modes
    case rgb8 = 2
    case gray8 = 3
}

// MARK: - Camera source protocol

/// Abstract camera source — the engine consumes frames from any source.
/// Implementations handle platform-specific camera APIs.
///
/// Known implementations:
/// - PhoneCameraSource: AVCaptureSession (iPhone/iPad)
/// - ARCameraSource: ARSession (ARKit on iOS/iPadOS)
/// - PassthroughCameraSource: Enterprise passthrough (Vision Pro)
/// - QuestCameraSource: Passthrough API (Meta Quest — Android side)
public protocol OpenEyesCameraSource: AnyObject {
    var cameraId: UInt32 { get }
    func start()
    func stop()
}

// MARK: - Engine wrapper

public final class OpenEyesEngine {
    private var engine: OpaquePointer?

    /// Motion magnitude (0.0 = static, higher = more motion)
    public private(set) var motionMagnitude: Float = 0.0

    /// Total frames processed
    public var frameCount: UInt64 {
        guard let engine = engine else { return 0 }
        return oe_get_frame_count(engine)
    }

    public init() {
        engine = oe_create()
    }

    deinit {
        if let engine = engine {
            oe_destroy(engine)
        }
    }

    // MARK: - Frame input

    /// Push a raw camera frame. Called from the camera callback — must be fast.
    ///
    /// On iOS: CVPixelBuffer from AVCaptureVideoDataOutput or ARFrame.capturedImage
    /// On visionOS: same path via enterprise passthrough API
    ///
    /// This is the HOT PATH. No Swift allocations here. Pointer goes straight to Rust.
    public func pushFrame(cameraId: UInt32, pixelBuffer: CVPixelBuffer, rotation: Int32 = 0) {
        guard let engine = engine else { return }

        CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly) }

        let width = UInt32(CVPixelBufferGetWidth(pixelBuffer))
        let height = UInt32(CVPixelBufferGetHeight(pixelBuffer))

        // Determine pixel format from the CVPixelBuffer's actual format
        let osType = CVPixelBufferGetPixelFormatType(pixelBuffer)
        let format: UInt32
        switch osType {
        case kCVPixelFormatType_32BGRA:
            format = OEPixelFormat.bgra8.rawValue
        case kCVPixelFormatType_420YpCbCr8BiPlanarVideoRange,
             kCVPixelFormatType_420YpCbCr8BiPlanarFullRange:
            format = OEPixelFormat.yuv420.rawValue
        default:
            format = OEPixelFormat.bgra8.rawValue // best guess
        }

        guard let baseAddress = CVPixelBufferGetBaseAddress(pixelBuffer) else { return }
        let dataSize = CVPixelBufferGetDataSize(pixelBuffer)
        let ptr = baseAddress.assumingMemoryBound(to: UInt8.self)

        // Straight to Rust. No copy. No Swift processing.
        oe_push_frame(engine, cameraId, ptr, dataSize, width, height, format, rotation)

        // Update cached motion (single float read, no allocation)
        motionMagnitude = oe_get_motion(engine)
    }

    // MARK: - AR Pose input

    /// Set camera pose from ARKit (ARFrame.camera.transform).
    /// Column-major 4x4 matrix — exactly what ARKit gives you.
    public func setCameraPose(cameraId: UInt32, transform: simd_float4x4) {
        guard let engine = engine else { return }

        // simd_float4x4 is column-major, which is what Rust expects.
        // Pass the raw 16 floats directly — no conversion needed.
        withUnsafePointer(to: transform) { ptr in
            ptr.withMemoryRebound(to: Float.self, capacity: 16) { floatPtr in
                oe_set_camera_pose(engine, cameraId, floatPtr)
            }
        }
    }

    // MARK: - Camera intrinsics

    /// Set camera intrinsics. Call once per camera, or when camera config changes.
    /// On ARKit: extract from ARFrame.camera.intrinsics
    public func setIntrinsics(
        cameraId: UInt32,
        fx: Double, fy: Double,
        cx: Double, cy: Double,
        width: UInt32, height: UInt32,
        k1: Double = 0, k2: Double = 0, k3: Double = 0
    ) {
        guard let engine = engine else { return }
        oe_set_intrinsics(engine, cameraId, fx, fy, cx, cy, width, height, k1, k2, k3)
    }
}

// MARK: - AVCaptureSession Camera Source (iPhone / iPad)

/// Camera source using AVCaptureSession — the standard iOS camera API.
/// Delivers CVPixelBuffer directly to the engine. Zero-copy on the hot path.
///
/// Also works on Mac Catalyst and macOS with minor config changes.
public final class PhoneCameraSource: NSObject, OpenEyesCameraSource, AVCaptureVideoDataOutputSampleBufferDelegate {
    public let cameraId: UInt32
    private let engine: OpenEyesEngine
    private let session = AVCaptureSession()
    private let outputQueue = DispatchQueue(label: "com.cambriantech.open-eyes.camera", qos: .userInteractive)

    public init(cameraId: UInt32, engine: OpenEyesEngine) {
        self.cameraId = cameraId
        self.engine = engine
        super.init()
        configureCaptureSession()
    }

    private func configureCaptureSession() {
        session.sessionPreset = .hd1280x720

        guard let device = AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .back),
              let input = try? AVCaptureDeviceInput(device: device) else { return }

        if session.canAddInput(input) {
            session.addInput(input)
        }

        let output = AVCaptureVideoDataOutput()
        // BGRA — matches OEPixelFormat::Bgra8, no conversion needed
        output.videoSettings = [
            kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA
        ]
        output.alwaysDiscardsLateVideoFrames = true
        output.setSampleBufferDelegate(self, queue: outputQueue)

        if session.canAddOutput(output) {
            session.addOutput(output)
        }
    }

    public func start() {
        session.startRunning()
    }

    public func stop() {
        session.stopRunning()
    }

    // AVCaptureVideoDataOutputSampleBufferDelegate — the hot path
    public func captureOutput(
        _ output: AVCaptureOutput,
        didOutput sampleBuffer: CMSampleBuffer,
        from connection: AVCaptureConnection
    ) {
        guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else { return }
        // Straight to Rust. CVPixelBuffer → oe_push_frame. No intermediate processing.
        engine.pushFrame(cameraId: cameraId, pixelBuffer: pixelBuffer)
    }
}

// MARK: - Future camera sources (stubs for architecture reference)

// ARCameraSource: Uses ARSession.delegate to get ARFrame.capturedImage (CVPixelBuffer)
// + ARFrame.camera.transform (simd_float4x4 pose) + ARFrame.camera.intrinsics.
// Same pushFrame + setCameraPose calls. ARKit gives us everything.
//
// PassthroughCameraSource (Vision Pro): Enterprise API gives passthrough camera access.
// visionOS ARKit provides spatial tracking. Same pattern — frames + poses to engine.
// The engine doesn't know it's running on a headset. It just gets frames + transforms.
//
// QuestCameraSource: See the Android binding — Meta Quest runs Android.
// Passthrough API gives camera frames. OVR SDK gives headset pose.
// Same oe_push_frame + oe_set_camera_pose calls via JNI.
